File: crypto-ios/AES_crypto/ak_aes_crypto.c
//
//  ak_aes_crypto.c
//  crypto-ios
//
//  Created by Andrew Kupetskiy on 11/23/17.
//  Copyright © 2017 AndrewK. All rights reserved.
//

#include <stdio.h>

#include "ak_aes_crypto.h"
#include "modes.h"
#include "handy.h"
#include "tassert.h"

void append_counter_to_nonce(uint8_t *nonce, uint32_t counter) {
    uint8_t ctr_array[4];
    ctr_array[0] = counter >> 24;
    ctr_array[1] = counter >> 16;
    ctr_array[2] = counter >>  8;
    ctr_array[3] = counter;
    
    memcpy(nonce, ctr_array, 4);
}

//-----------------------------------------------------------------

void ak_aes_init(ak_aes_context *ctx,
                 const uint8_t *shared_secret,
                 size_t shared_secret_size) {
    memset(ctx, 0, sizeof *ctx);
    ctx->counter = 0;
    
    // init nonce from shared secret
    // nonce = shared_secret[19...31];
    memcpy(ctx->nonce, shared_secret+19, NONCE_SIZE);
    
    //then apply counter value onto it
    append_counter_to_nonce(ctx->nonce, ctx->counter);
        
    // init basic aes crypto
    cf_aes_context *ccm_context = (cf_aes_context *)malloc(sizeof(cf_aes_context));
    
    // pass secret key to it
    uint8_t secret_key[SECRET_KEY_SIZE];
    memcpy(secret_key, shared_secret, SECRET_KEY_SIZE);
    
    cf_aes_init(ccm_context, secret_key, SECRET_KEY_SIZE);
    
    ctx->ccm_ctx = ccm_context;
}

void ak_aes_finish(ak_aes_context *ctx)
{
    free(ctx->ccm_ctx);
    mem_clean(ctx, sizeof *ctx);
}

//----------------------------------------------------------------------

void ak_aes_encrypt(ak_aes_context *ctx,
                    const uint8_t *plain_data, size_t plain_data_size,
                    uint8_t *encrypted_data,
                    uint8_t *tag, size_t ntag) {
    
    //apply counter value onto it
    append_counter_to_nonce(ctx->nonce, ctx->counter);
    
    cf_ccm_encrypt(&cf_aes, ctx->ccm_ctx,
                   plain_data, plain_data_size, L_SIZE,
                   NULL, 0,
                   ctx->nonce, NONCE_SIZE,
                   encrypted_data,
                   tag, ntag);
    
    //increment counter
    ctx->counter = ctx->counter + 1;
}

int ak_aes_decrypt(ak_aes_context *ctx,
                   const uint8_t *encrypted_data, size_t encrypted_data_size,
                   const uint8_t *tag, size_t ntag,
                   uint8_t *plain_data) {
    
    int err = 0;
    err = cf_ccm_decrypt(&cf_aes, ctx->ccm_ctx,
                         encrypted_data, encrypted_data_size, L_SIZE,
                         NULL, 0,
                         ctx->nonce, NONCE_SIZE,
                         tag, ntag,
                         plain_data);
    
    //increment counter
    ctx->counter = ctx->counter + 1;
    
    return err;
}

// -------------------------------------------------------------------

void ak_aes_encrypt_and_pack(ak_aes_context *ctx,
                             const uint8_t *plain_data, size_t plain_data_size,
                             uint8_t *package, uint8_t *package_size) {
    uint8_t encrypted_data[plain_data_size], tag[TAG_SIZE];
    //apply counter value onto it
    
    ak_aes_encrypt(ctx, plain_data, plain_data_size, encrypted_data, tag, TAG_SIZE);
    
    // copy tag first
    memcpy(package, tag, TAG_SIZE);
    // then message
    memcpy(package + TAG_SIZE, encrypted_data, plain_data_size);
}

int ak_aes_unpack_and_decrypt(ak_aes_context *ctx,
                              const uint8_t *package, size_t package_size,
                              uint8_t *plain_data) {
    size_t encrypted_data_size = package_size - TAG_SIZE;
    uint8_t encrypted_data[encrypted_data_size], tag[TAG_SIZE];
    
    // copy tag first
    memcpy(tag, package, TAG_SIZE);
    // then message
    memcpy(encrypted_data,package + TAG_SIZE, encrypted_data_size);
    
    return ak_aes_decrypt(ctx, encrypted_data, encrypted_data_size, tag, TAG_SIZE, plain_data);
}

// --------------------------------------------------------------

File: crypto-ios/AES_crypto/AKAESCrypto.m
//
//  AKAESCrypto.m
//  crypto-ios
//
//  Created by Andrew Kupetskiy on 11/23/17.
//  Copyright © 2017 AndrewK. All rights reserved.
//

#import "AKAESCrypto.h"
#import "ak_aes_crypto.h"

@interface AKAESCrypto ()
@property (nonatomic, strong) NSData *secret;
@property (nonatomic, assign) ak_aes_context *aes_ccm_ctx;
@end

@implementation AKAESCrypto

- (instancetype)initWithSecret:(NSData *)secret {
    if ([secret length] != 32) {
        return nil;
    }
    
    if (self = [super init]) {
        _secret = secret;
        
        _aes_ccm_ctx = malloc(sizeof(ak_aes_context));
        ak_aes_init(_aes_ccm_ctx, [secret bytes], [secret length]);
    }
    
    return self;
}

- (void)dealloc {
    free(_aes_ccm_ctx);
}

- (NSData *)encryptWithData:(NSData *)data {
    size_t package_len = data.length + TAG_SIZE;
    uint8_t package[package_len];
    
    // magic in C func
    ak_aes_encrypt_and_pack(self.aes_ccm_ctx, data.bytes, data.length,package, package);
    
    // convert and send
    return [[NSData alloc] initWithBytes:package length:package_len];
}

- (NSData *)decryptWithData:(NSData *)encryptedData {
    size_t plain_data_len = encryptedData.length - TAG_SIZE;
    uint8_t plain_data[plain_data_len];
    
    // magic in C func
    int error = ak_aes_unpack_and_decrypt(self.aes_ccm_ctx, encryptedData.bytes, encryptedData.length, plain_data);
    
    if (error) {
        return nil;
    }
    // convert and send
    return [[NSData alloc] initWithBytes:plain_data length:plain_data_len];
    
    return nil;
}

+ (void)test_pack_unpack {
    uint8_t cipher[12+TAG_SIZE], decrypted[12];
    
    void *secret = "SxqlOW3UadPQMlo3pOVloA8Bk2MY9WYI";
    void *msg = "hello world\0";
    
    ak_aes_context ctx;
    ak_aes_init(&ctx, secret, 32);
    
    ak_aes_encrypt_and_pack(&ctx, msg, 12, cipher, 12 + TAG_SIZE);
    
    int err;
    err = ak_aes_unpack_and_decrypt(&ctx, cipher, 12+TAG_SIZE, decrypted);
    
    assert(err == 0);
    assert(memcmp(decrypted, msg, 12) == 0);
    
    cipher[0] ^= 0xff;
    err = ak_aes_unpack_and_decrypt(&ctx, cipher, 12+TAG_SIZE, decrypted);
    assert(err == 1);
    
}

+ (void)test_crypto_C_realisation {
    uint8_t cipher[12], tag[16], decrypted[12];
    
    void *secret = "SxqlOW3UadPQMlo3pOVloA8Bk2MY9WYI";
    void *msg = "hello world\0";
    
    ak_aes_context ctx;
    ak_aes_init(&ctx, secret, 32);
    
    ak_aes_encrypt(&ctx, msg, 12, cipher, tag, 16);
    
    int err;
    err = ak_aes_decrypt(&ctx, cipher, 12, tag, 16, decrypted);
    
    assert(err == 0);
    assert(memcmp(decrypted, msg, 12) == 0);
    
    tag[0] ^= 0xff;
    err = ak_aes_decrypt(&ctx, cipher, 12, tag, 16, decrypted);
    assert(err == 1);
    
    [self test_pack_unpack];
}

@end
